src/librustc/traits/error_reporting.rs

```diff
@@ -1236,60 +1236,109 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
+        // evaluate to a type that *would* satisfy the trait binding. If it would, suggest calling
+        // it: `bar(foo)` → `bar(foo())`. This case is *very* likely to be hit if `foo` is `async`.
+
+        let new_trait_ref = ty::TraitRef {
+            def_id: trait_ref.def_id(),
+            substs: self.tcx.mk_substs_trait(output_ty.skip_binder(), &[]),
+        };
+        let obligation = Obligation::new(
+            obligation.cause.clone(),
+            obligation.param_env,
+            new_trait_ref.to_predicate(),
+        );
+        let get_name = |err: &mut DiagnosticBuilder<'_>, kind: &hir::PatKind| -> Option<String> {
+            // Get the local name of this closure. This can be inaccurate because
+            // of the possibility of reassignment, but this should be good enough.
+            match &kind {
+                hir::PatKind::Binding(hir::BindingAnnotation::Unannotated, _, name, None) => {
+                    Some(format!("{}", name))
+                }
+                _ => {
+                    err.note(&msg);
+                    None
+                }
+            }
+        };
+        match self.evaluate_obligation(&obligation) {
+            Ok(EvaluationResult::EvaluatedToOk) |
+            Ok(EvaluationResult::EvaluatedToOkModuloRegions) |
+            Ok(EvaluationResult::EvaluatedToAmbig) => {
+                let hir = self.tcx.hir();
+                // Get the name of the callable and the arguments to be used in the suggestion.
+                let snippet = match hir.get_if_local(def_id) {
+                    Some(hir::Node::Expr(hir::Expr {
+                        kind: hir::ExprKind::Closure(_, decl, _, span, ..),
+                        ..
+                    })) => {
+                        err.span_label(*span, "consider calling this closure");
+                        let hir_id = match hir.as_local_hir_id(def_id) {
+                            Some(hir_id) => hir_id,
+                            None => return,
+                        };
+                        let parent_node = hir.get_parent_node(hir_id);
+                        let name = match hir.find(parent_node) {
+                            Some(hir::Node::Stmt(hir::Stmt {
+                                kind: hir::StmtKind::Local(local), ..
+                            })) => match get_name(err, &local.pat.kind) {
+                                Some(name) => name,
+                                None => return,
+                            },
+                            // Different to previous arm because one is `&hir::Local` and the other
+                            // is `P<hir::Local>`.
+                            Some(hir::Node::Local(local)) => match get_name(err, &local.pat.kind) {
+                                Some(name) => name,
+                                None => return,
+                            },
+                            _ => return,
+                        };
```

Comment:
This could be its own helper function `-> Option<_>` imo + doc comment. (also move the closure in)